Kubevirt e2e ipamclaim

qinqon opened the pull request:
**- What this PR does and why is it needed**
kubevirt, e2e: Add persistent ips tests
To verify the ovn-kubernetes integration with IPAMClaim this change
add kubevirt VirtualMachines live migration and restart with
multi-homing and persistent ips tests.

qinqon commented on contrib/kind.sh:
```diff
@@ -1353,7 +1353,14 @@ function install_kubevirt() {
     if [ "$(kubectl get kubevirts -n kubevirt kubevirt -ojsonpath='{.status.phase}')" != "Deployed" ]; then
-      kubectl apply -f "${kubevirt_release_url}/kubevirt-operator.yaml"
+      local virt_controller_image=quay.io/ellorent/virt-controller:persistentips
+      local virt_handler_image=quay.io/ellorent/virt-handler:persistentips
```
This is only use to test https://github.com/kubevirt/kubevirt/pull/11410 when that is merged we will use kubevirt nightlies.

qinqon replied:
Humm... I will recheck, maybe virt-handler was just copy paste.

maiqueb commented on test/e2e/kubevirt.go:
```diff
@@ -721,159 +813,338 @@ passwd:
+			expectedAddreses = waitVirtualMachineAddressesFromStatus(vmi)
+
+			td.test.cmd()
+			time.Sleep(time.Second)
```
I don't get this sleep, but it does not look good. No sleep does.

qinqon replied:
So after a restart/live-migration we have to wait a little for the vmi to catch up with new state, we can try to use a `Eventually` with expected NotReady or the like but I don't think is it worth it.

maiqueb commented on test/e2e/kubevirt.go:
```diff
@@ -524,7 +499,104 @@ passwd:
+			return addresses
+		}
+
+		waitVirtualMachineAddressesFromStatus = func(vmi *kubevirtv1.VirtualMachineInstance, expectedNumberOfAddresses int) []string {
```
would it make sense instead to pass a variadic number of subnets and for each of them assert there's an IP in the VM status in that range ?

qinqon replied:
well the VMI we are configuring only expose the interface from the nad, the fact that those IPs belong to the nad.subnet is already tested at multihoming, what we want to test here is that the IPs whatever they are should prevail.

maiqueb replied:
yes, but by doing that, the function's signature looks a lot better IMO.

maiqueb commented on test/e2e/kubevirt.go:
```diff
@@ -524,7 +499,104 @@ passwd:
+					return nil, err
+				}
+				addresses := []string{}
+				for _, iface := range vmi.Status.Interfaces {
```
I'm assuming we don't report ipv6 link-local addresses. Right ?

qinqon replied:
nope

qinqon commented:
persistentips is crashing
```
I0315 13:01:37.117352   17544 obj_retry.go:713] Delete event received for *v1alpha1.IPAMClaim kv-live-migration-8064/worker1.net1
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x1a79ca1]

goroutine 104 [running]:
github.com/ovn-org/ovn-kubernetes/go-controller/pkg/allocator/persistentips.(*Allocator).Delete(0x1537604?, 0xc000208030?)
	/home/ellorent/Documents/cnv/upstream/ovn-kubernetes/go-controller/pkg/allocator/persistentips/allocator.go:103 +0x21
github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn.(*BaseSecondaryNetworkController).DeleteSecondaryNetworkResourceCommon(0xc000978340, {0x2712e08?, 0x22d7b40}, {0x22d7b40?, 0xc00023d680?}, {0x0?, 0x0?})
	/home/ellorent/Documents/cnv/upstream/ovn-kubernetes/go-controller/pkg/ovn/base_network_controller_secondary.go:197 +0x4bc
github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn.(*baseSecondaryLayer2NetworkControllerEventHandler).DeleteResource(0xc0008a3e30?, {0x22d7b40?, 0xc00023d680?}, {0x0?, 0x0?})
	/home/ellorent/Documents/cnv/upstream/ovn-kubernetes/go-controller/pkg/ovn/base_secondary_layer2_network_controller.go:134 +0x169
github.com/ovn-org/ovn-kubernetes/go-controller/pkg/retry.(*RetryFramework).WatchResourceFiltered.func3.1({0xc0010fe630, 0x23})
	/home/ellorent/Documents/cnv/upstream/ovn-kubernetes/go-controller/pkg/retry/obj_retry.go:731 +0xdf
github.com/ovn-org/ovn-kubernetes/go-controller/pkg/retry.(*RetryFramework).DoWithLock(0xc001110b90, {0xc0010fe630, 0x23}, 0xc000199cb8)
	/home/ellorent/Documents/cnv/upstream/ovn-kubernetes/go-controller/pkg/retry/obj_retry.go:137 +0xbe
github.com/ovn-org/ovn-kubernetes/go-controller/pkg/retry.(*RetryFramework).WatchResourceFiltered.func3({0x22d7b40, 0xc00023d680})

```

maiqueb commented:
> persistentips is crashing
> 
> ```
> I0315 13:01:37.117352   17544 obj_retry.go:713] Delete event received for *v1alpha1.IPAMClaim kv-live-migration-8064/worker1.net1
> panic: runtime error: invalid memory address or nil pointer dereference
> [signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x1a79ca1]
> 
> goroutine 104 [running]:
> github.com/ovn-org/ovn-kubernetes/go-controller/pkg/allocator/persistentips.(*Allocator).Delete(0x1537604?, 0xc000208030?)
> 	/home/ellorent/Documents/cnv/upstream/ovn-kubernetes/go-controller/pkg/allocator/persistentips/allocator.go:103 +0x21
> github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn.(*BaseSecondaryNetworkController).DeleteSecondaryNetworkResourceCommon(0xc000978340, {0x2712e08?, 0x22d7b40}, {0x22d7b40?, 0xc00023d680?}, {0x0?, 0x0?})
> 	/home/ellorent/Documents/cnv/upstream/ovn-kubernetes/go-controller/pkg/ovn/base_network_controller_secondary.go:197 +0x4bc
> github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn.(*baseSecondaryLayer2NetworkControllerEventHandler).DeleteResource(0xc0008a3e30?, {0x22d7b40?, 0xc00023d680?}, {0x0?, 0x0?})
> 	/home/ellorent/Documents/cnv/upstream/ovn-kubernetes/go-controller/pkg/ovn/base_secondary_layer2_network_controller.go:134 +0x169
> github.com/ovn-org/ovn-kubernetes/go-controller/pkg/retry.(*RetryFramework).WatchResourceFiltered.func3.1({0xc0010fe630, 0x23})
> 	/home/ellorent/Documents/cnv/upstream/ovn-kubernetes/go-controller/pkg/retry/obj_retry.go:731 +0xdf
> github.com/ovn-org/ovn-kubernetes/go-controller/pkg/retry.(*RetryFramework).DoWithLock(0xc001110b90, {0xc0010fe630, 0x23}, 0xc000199cb8)
> 	/home/ellorent/Documents/cnv/upstream/ovn-kubernetes/go-controller/pkg/retry/obj_retry.go:137 +0xbe
> github.com/ovn-org/ovn-kubernetes/go-controller/pkg/retry.(*RetryFramework).WatchResourceFiltered.func3({0x22d7b40, 0xc00023d680})
> ```

We had two event handlers listening to `IPAMClaim`s deletes. 

We need to ensure when the VMs are deleted (which in turn deletes the `IPAMClaim`s) the OVN-K pods don't fumble. 

Does this make sense ?

maiqueb commented on test/e2e/kubevirt.go:
```diff
@@ -347,6 +262,59 @@ passwd:
+		}
+
+		checkPodHasIPsAtNetwork = func(netName string, expectedNumberOfAddresses int) func(Gomega, *corev1.Pod) {
+			return func(g Gomega, pod *corev1.Pod) {
```
having to receive gomega is avoidable; I would rather have this function return an error, and assert in the caller.

qinqon replied:
This is better since you will get a proper matching error 
```golang
g.Expect(err).ToNot(HaveOccurred())                             
g.Expect(netStatus).To(HaveLen(1))                              
g.Expect(netStatus[0].IPs).To(HaveLen(expectedNumberOfAddresses))
```

Instead of a custom error you have to manually compose to be descriptive of the type of failure.

maiqueb replied:
might be just me, but I unfortunately don't follow why what you're proposing is better.

maiqueb commented on test/e2e/kubevirt.go:
```diff
@@ -540,7 +526,104 @@ passwd:
+				if err != nil {
+					return nil, err
+				}
+				addresses := []string{}
```
nit: `var addresses []string` instead.

maiqueb commented on test/e2e/kubevirt.go:
the majority of this commit is a refactor to re-use some code later on. 

can we split it in two commits, so we get a commit with just the new tests ? 

Having to review everything in one dilutes the changes, and makes it harder to focus on the important bits.

maiqueb replied:
I am unsure if you acted on this ... 

I still see the topmost commit with many, many changes, which impact more than the new tests.

maiqueb commented on test/e2e/kubevirt.go:
```diff
@@ -540,7 +526,104 @@ passwd:
+			return addresses
+		}
+
+		waitVirtualMachineAddressesFromStatus = func(vmi *kubevirtv1.VirtualMachineInstance, expectedNumberOfAddresses int) []string {
```
I don't understand why a function named wait for <something> returns a list of addresses. 

IMO, that's a clear indication the function does more than one thing. Consider splitting it in two.

qinqon replied:
Renaming to virtualMachineAddressesFromStatus

maiqueb commented on test/e2e/kubevirt.go:
```diff
@@ -540,7 +526,104 @@ passwd:
 		}
-		fcosVM = func(idx int, labels map[string]string, butane string) (*kubevirtv1.VirtualMachine, error) {
+
+		addressesFromStatus = func(vmi *kubevirtv1.VirtualMachineInstance) func() ([]string, error) {
```
why does this return a function ? Can't you just return the addresses and error, and wrap it in a func call whenever required ? (e.g. eventually / consistently blocks )

IMO it just makes the code slightly more complex.

qinqon replied:
It allow you to use Eventually without a closure 
```golang
Eventually(addressesFromStatus(vmi)).                               
                WithPolling(time.Second).                                       
                WithTimeout(10*time.Second).                                    
                Should(HaveLen(expectedNumberOfAddresses), step)       
```

maiqueb replied:
it's probably just me, but I prefer to have that shoved in my face since it enforces a good practice of knowing I'm passing it a function. 

This works, but I need to keep on my toes and double check every call out there. If I see the closure, I know it's correct. This way, I don't. 

... but this is opinionated.

maiqueb commented on test/e2e/kubevirt.go:
```diff
@@ -540,7 +526,104 @@ passwd:
+		createVirtualMachine = func(vm *kubevirtv1.VirtualMachine) {
+			By(fmt.Sprintf("Create virtual machine %s", vm.Name))
+			vmCreationRetries := 0
+			Eventually(func() error {
```
shouldn't this call have an offset ?

maiqueb commented on test/e2e/kubevirt.go:
```diff
@@ -540,7 +526,104 @@ passwd:
+		createVirtualMachineInstance = func(vmi *kubevirtv1.VirtualMachineInstance) {
+			By(fmt.Sprintf("Create virtual machine instance %s", vmi.Name))
+			vmiCreationRetries := 0
+			Eventually(func() error {
```
ditto (offset)

maiqueb reviewed (commented):
I would prefer if you could split the new tests into a dedicated commit; that would make it easier to review.

Also, I think most of the functions you are adding could avoid using closures. What do we get by using them ? IMHO it is just harder to review, since we need to pay extra attention to the shared variable scope.

qinqon commented:
rebased after ipamclaim feature merged.

qinqon commented:
> I would prefer if you could split the new tests into a dedicated commit; that would make it easier to review.
> 
ok

> Also, I think most of the functions you are adding could avoid using closures. What do we get by using them ? IMHO it is just harder to review, since we need to pay extra attention to the shared variable scope.

Closure are the proper way at ginkgo to make functions private to the test spec, this prevent other test using your functions just because they are there.

qinqon commented:
@maiqueb I have split the PR at multiple commits as you suggested.

qinqon commented on test/e2e/kubevirt.go:
```diff
@@ -66,54 +80,12 @@ var _ = Describe("Kubevirt Virtual Machines", func() {
 		crClient           crclient.Client
 		namespace          string
 		tcpServerPort      = int32(9900)
-		isDualStack        = false
```
I have split the PR.

qinqon commented on test/e2e/kubevirt.go:
```diff
@@ -66,54 +80,12 @@ var _ = Describe("Kubevirt Virtual Machines", func() {
 		clientSet          kubernetes.Interface
 		// Systemd resolvd prevent resolving kube api service by fqdn, so
 		// we replace it here with NetworkManager
-		butane = fmt.Sprintf(`
```
I have split the PR

oshoval commented on test/e2e/kubevirt.go:
```diff
@@ -742,160 +872,400 @@ passwd:
+			}),
+		)
 	})
+	Context("with non default network and persistent ips configured", func() {
```
is there a string like "persistent ips" so we can use
WHAT=".*persistent ip.*" to run all and just persistent ips tests ?
if not please consider adding, it can be helpful

maiqueb commented:
@qinqon the tests are not passing. 

Can you look into this ?

maiqueb commented on test/e2e/kubevirt.go:
```diff
@@ -870,160 +872,400 @@ passwd:
+
+		AfterEach(func() {
+			for _, node := range selectedNodes {
+				unlabelNode(node.Name, namespace)
```
please handle the error here.

maiqueb reviewed (changes requested):
You're missing the `WithOffset` / `ExpectWithOffset` in many of your helpers.

maiqueb commented:
@qinqon there's also a failing live-migration test (on the default cluster network).

Please check that out. I've sorted out the secondary ones in #4349 .

qinqon commented:
> @qinqon there's also a failing live-migration test (on the default cluster network).
> 
> Please check that out. I've sorted out the secondary ones in #4349 .

Taking your fixes from there.

qinqon commented:
> @qinqon there's also a failing live-migration test (on the default cluster network).
> 

Right, this is failing locally too

```
 [FAILED] [143.514 seconds]
Kubevirt Virtual Machines with default pod network when live migration [It] with pre-copy succeeds, should keep connectivity
/home/ellorent/Documents/cnv/upstream/ovn-kubernetes/test/e2e/kubevirt.go:1082

  [FAILED] Timed out after 12.004s.
  worker1: after live migrate for the first time: Check e/w tcp traffic: testpod-ovn-worker: 
  Expected success, but got an error:
      <*errors.errorString | 0xc000d140d0>: 
      failed to run [curl http://10.244.2.20:8000] at VMI worker1, error: expect: timer expired after 10 seconds
      {
          s: "failed to run [curl http://10.244.2.20:8000] at VMI worker1, error: expect: timer expired after 10 seconds",
      }
  In [It] at: /home/ellorent/Documents/cnv/upstream/ovn-kubernetes/test/e2e/kubevirt.go:344 @ 05/16/24 11:42:22.389

  There were additional failures detected.  To view them in detail run ginkgo -vv
```

maiqueb commented on test/e2e/kubevirt.go:
```diff
@@ -353,6 +268,59 @@ passwd:
+					output := ""
+					Eventually(func() error {
+						var err error
+						output, err = kubevirt.RunCommand(vmi, fmt.Sprintf("curl http://%s", net.JoinHostPort(podIP, "8000")), polling)
```
you're again passing polling. If you check the `kubevirt.RunCommand` function, it is supposed to receive the timeout, not the polling interval:
https://github.com/ovn-org/ovn-kubernetes/blob/51b2596f711a182b0f09182d30c4321529671d43/test/e2e/kubevirt/console.go#L84

qinqon replied:
I am leaving this exactly as it is at `origin/master` make sense to pass polling so we "break" RunCommand inside eventually and it retries, if we don't do that it will not retry (maybe curl is already retrying, don't know).

maiqueb replied:
it's very different to wait for 15 seconds than wait for 1 minute ... 

anyway ... whatever. if you want to wait for 1 second, and it just works, it's fine by me.

qinqon replied:
we wait for 1 minute at Eventually not at RunCommand.

qinqon commented:
Moving these test to https://github.com/ovn-org/ovn-kubernetes/pull/4349

qinqon commented:
PR re-open to use ipamclaim controller instead of kubevirt wip.

jcaamano commented on test/e2e/kubevirt.go:
```diff
@@ -870,160 +876,402 @@ passwd:
+			}),
+		)
+	})
+	Context("with non default network and persistent ips configured", func() {
```
is mthis more rather `with secondary network...`, like the default network is still there and is also live migrated as in the previous spec?

maiqueb replied:
this is a good point: we could pin-point the feature and just drop the cluster default network from the VM. i.e. only have a single secondary network interface.

qinqon replied:
I will reword it as "with secondary network"

maiqueb replied:
and how about dropping the default cluster network attachment from the VM's interface list ? 

That has nothing to do with the feature.
@qinqon

maiqueb replied:
My bad then. All good.

jcaamano commented on test/e2e/kubevirt.go:
```diff
@@ -870,160 +876,402 @@ passwd:
-	if len(unstructuredVMIMigrations.Items) == 0 {
-		return nil, fmt.Errorf("empty migration list")
-	}
+			step := by(vmi.Name, "Login to virtual machine for the first time")
```
not here, but in some other places not changed by this PR  you have a unnecessary `fmt.Sprintf` on your `by` statements

qinqon replied:
Will remove Sprintf from were is not really needed.

jcaamano commented on test/e2e/kubevirt.go:
```diff
@@ -870,160 +876,402 @@ passwd:
+
+			checkEastWestTraffic(vmi, testPodsIPs, step)
+
+			td.test.cmd()
```
Missing a `by` statement for this step

jcaamano commented on test/e2e/kubevirt.go:
```diff
@@ -870,160 +876,402 @@ passwd:
+				test:     liveMigrate,
+				topology: "layer2",
+			}),
+			Entry(nil, testData{
+				resource: virtualMachineInstance,
+				test:     liveMigrate,
+				topology: "localnet",
+			}),
+			Entry(nil, testData{
+				resource: virtualMachineInstance,
+				test:     liveMigrate,
+				topology: "layer2",
+			}),
```
What's the point of these two? Or why don't we have restart test for VMIs? Out of curiosity.

qinqon replied:
You cannot restart VMIs, you can only restart VMs, as far as I know we decided to also support VMIs.

jcaamano reviewed (commented):
no major issues on my part, just nits

qinqon commented on test/e2e/kubevirt.go:
```diff
@@ -133,81 +103,24 @@ passwd:
-					break // we want just three of them
+		isDualStack = func() bool {
+			nodeList, err := fr.ClientSet.CoreV1().Nodes().List(context.Background(), metav1.ListOptions{})
+			Expect(err).WithOffset(1).ToNot(HaveOccurred())
```
Quite interesting, I didn't know about `GinkgoHelper`

qinqon commented:
@maiqueb @jcaamano I have squashed the last two commits (adding e2e tests and configuring underlay) to simplify stuff, also addressed @jcaamano comments.
